tools/autograd/gen_variable_type.py

```diff
@@ -851,7 +851,14 @@ def find_args_with_derivatives(
         and (len(differentiable_outputs) > 0)
     )
 
-    if info is not None and info.has_derivatives and not requires_derivative:
+    # TODO: check with Alban/Jeffrey.
+    # Now that we have out= ops that can have no returns, how should this work?
```

Comment:
out= ops should never end up here.
Also having no returns should not be a problem in theory (even it might not work in practice if it is not used today).

Reply:
out= ops should never end up here? I was playing around with it / printing stuff out, and it looks like a bunch of out= ops go down this codegen path- most of them get a valid `DifferentiabilityInfo` object because of [this code](https://github.com/pytorch/pytorch/blob/b2507592424f4a14ab6613cd05b20f18f49f2663/torchgen/api/autograd.py#L319) - if `mul.out` doesn't have a derivative, we find the one for `mul.Tensor`.

The out= ops go down this path, but don't trip up the AssertionError below, because `requires_derivative` is true (because `len(differentiable_inputs) > 0 and len(differentiable_inputs) > 0`).

But now that we have some out= ops that have no returns (like `split_copy.Tensor_out`), then they'll report `len(differentiable_outputs) == 0`, and `requires_derivative` will be false, which trips the below assert.

That only broke in this PR because we used to not properly pair up the existing no-return out= ops with their functional variants - but now we do.

It looks like the existing out= ops don't trip up here because `len(differentiable_inputs)` and `len(differentiable_outputs)` are both `> 0`. But for the